### ISSUE TITLE
Flaky-test: KeySharedSubscriptionTest.testRemoveFirstConsumer

### ISSUE THREAD

**Issue by lhotari** [apache/pulsar#11426]
KeySharedSubscriptionTest is flaky. The testRemoveFirstConsumer test method fails sporadically.

[example failure](https://github.com/apache/pulsar/pull/11343/checks?check_run_id=3130775265#step:10:1061)

```
  Error:  Tests run: 46, Failures: 1, Errors: 0, Skipped: 19, Time elapsed: 204.45 s <<< FAILURE! - in org.apache.pulsar.client.api.KeySharedSubscriptionTest
  Error:  testRemoveFirstConsumer(org.apache.pulsar.client.api.KeySharedSubscriptionTest)  Time elapsed: 0.228 s  <<< FAILURE!
  java.lang.AssertionError: expected [0] but found [10]
  	at org.testng.Assert.fail(Assert.java:99)
  	at org.testng.Assert.failNotEquals(Assert.java:1037)
  	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
  	at org.testng.Assert.assertEquals(Assert.java:122)
  	at org.testng.Assert.assertEquals(Assert.java:907)
  	at org.testng.Assert.assertEquals(Assert.java:917)
  	at org.apache.pulsar.client.api.KeySharedSubscriptionTest.testRemoveFirstConsumer(KeySharedSubscriptionTest.java:685)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
  	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
  	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
  	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
  	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
  	at java.base/java.lang.Thread.run(Thread.java:829)
  
```

**Comment by lhotari** [comment]
@Jason918 The problem isn't resolved. I'm reopening this issue. Here's a recent failure: https://github.com/apache/pulsar/runs/4236580815?check_suite_focus=true#step:9:3246 . It's a test run for PR #12853 and it included your attempt #12807 to fix the flaky test.

```
Error:  Tests run: 50, Failures: 1, Errors: 0, Skipped: 23, Time elapsed: 221.229 s <<< FAILURE! - in org.apache.pulsar.client.api.KeySharedSubscriptionTest
  Error:  testRemoveFirstConsumer(org.apache.pulsar.client.api.KeySharedSubscriptionTest)  Time elapsed: 0.314 s  <<< FAILURE!
  java.lang.AssertionError: expected [0] but found [10]
  	at org.testng.Assert.fail(Assert.java:99)
  	at org.testng.Assert.failNotEquals(Assert.java:1037)
  	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
  	at org.testng.Assert.assertEquals(Assert.java:122)
  	at org.testng.Assert.assertEquals(Assert.java:907)
  	at org.testng.Assert.assertEquals(Assert.java:917)
  	at org.apache.pulsar.client.api.KeySharedSubscriptionTest.testRemoveFirstConsumer(KeySharedSubscriptionTest.java:698)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
  	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
  	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
  	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
  	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
  	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by Jason918** [comment]
Yes, I got reproduced locally. And I got a detailed DEBUG log. I will keep look into this.

**Comment by Jason918** [comment]
[testRemoveFirstConsumer.log](https://github.com/apache/pulsar/files/7555076/testRemoveFirstConsumer.log)

Debug level log here. 
Anyone familiar with KeyShare consuming mode could help.

**Comment by Jason918** [comment]
I have located this issue, and the test case is actually ok, but it caused by a  race condition in `PersistentDispatcherMultipleConsumers`, which indeed would causing the unordered consuming.

One can reproduce it easily in local environment by changing the 100ms to 1ms here.
https://github.com/apache/pulsar/blob/552360455382e33901b6d74c26f883ffb6907ba9/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentStickyKeyDispatcherMultipleConsumers.java#L294

I would submit an issue with detailed explanation about this.

**Comment by codelipenghui** [comment]
The issue had no activity for 30 days, mark with Stale label.

**Comment by Jason918** [comment]
It's fixed by #12890.
